<?xml version="1.0"?>
<robot name="common_base" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:property name="model_name" value="$(arg model_name)"/>

  <link name='chassis'>
    <pose>-0.151427 -0 0.175 0 -0 0</pose>
    <inertial>
      <mass value="1.14395"/>
      <inertia ixx="0.126164" ixy="0" ixz="0" iyy="0.416519" iyz="0" izz="0.481014"/>
    </inertial>
    <visual name='visual'>
      <geometry>
        <box size="2.01142 1 0.568726"/>
      </geometry>
    </visual>
    <collision name='collision'>
      <geometry>
        <box size="2.01142 1 0.568726"/>
      </geometry>
    </collision>
  </link>

  <link name="lidar_link">
    <pose>0 0 0.5 0 0 0</pose>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.000166667" ixy="0" ixz="0" iyy="0.000166667" iyz="0" izz="0.000166667"/>
    </inertial>
    <collision name="collision">
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual name="visual">
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </visual>
  </link>

  <link name='left_wheel'>
    <pose>0.554283 0.625029 -0.025 -1.5707 0 0</pose>
    <inertial>
      <mass value="2"/>
      <inertia ixx="0.145833" ixy="0" ixz="0" iyy="0.145833" iyz="0" izz="0.125"/>
    </inertial>
    <visual name='visual'>
      <geometry>
        <sphere radius="0.3"/>
      </geometry>
    </visual>
    <collision name='collision'>
      <geometry>
        <sphere radius="0.3"/>
      </geometry>
    </collision>
  </link>

  <link name='right_wheel'>
    <pose>0.554282 -0.625029 -0.025 -1.5707 0 0</pose>
    <inertial>
      <mass value="2"/>
      <inertia ixx="0.145833" ixy="0" ixz="0" iyy="0.145833" iyz="0" izz="0.125"/>
    </inertial>
    <visual name='visual'>
      <geometry>
        <sphere radius="0.3"/>
      </geometry>
    </visual>
    <collision name='collision'>
      <geometry>
        <sphere radius="0.3"/>
      </geometry>
    </collision>
  </link>


  <joint name='lidar_sensor_joint' type='fixed'>
    <parent link="chassis"/>
    <child link="lidar_link"/>
  </joint>

  <joint name='left_wheel_joint' type='revolute'>
    <parent link="chassis"/>
    <child link="left_wheel"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" velocity="1.0" lower="-1.79769e+308" upper="1.79769e+308"/>
  </joint>

  <joint name='right_wheel_joint' type='revolute'>
    <parent link="chassis"/>
    <child link="right_wheel"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" velocity="1.0" lower="-1.79769e+308" upper="1.79769e+308"/>
  </joint>


  <gazebo>
    <plugin
      filename="libignition-gazebo-diff-drive-system.so"
      name="ignition::gazebo::systems::DiffDrive"
    >
      <left_joint>left_wheel_joint</left_joint>
      <right_joint>right_wheel_joint</right_joint>
      <wheel_separation>1.25</wheel_separation>
      <wheel_radius>0.3</wheel_radius>
      <odom_publish_frequency>1</odom_publish_frequency>
    </plugin>

    <plugin filename="libignition-gazebo-logical-camera-system.so"
        name="ignition::gazebo::systems::LogicalCamera">
    </plugin>

    <plugin
      filename="ignition-gazebo-imu-system"
      name="ignition::gazebo::systems::Imu">
    </plugin>

    <plugin filename="libignition-gazebo-pose-publisher-system.so"
        name="ignition::gazebo::systems::PosePublisher">
        <publish_link_pose>true</publish_link_pose>
        <publish_sensor_pose>false</publish_sensor_pose>
        <publish_collision_pose>false</publish_collision_pose>
        <publish_visual_pose>false</publish_visual_pose>
        <publish_nested_model_pose>true</publish_nested_model_pose>
        <use_pose_vector_msg>true</use_pose_vector_msg>
        <static_publisher>true</static_publisher>
        <static_update_frequency>30</static_update_frequency>
    </plugin>
  </gazebo>

  <gazebo reference="chassis">
     <sensor name="imu_sensor" type="imu">
      <always_on>1</always_on>
      <update_rate>1</update_rate>
      <visualize>true</visualize>
      <topic>model/${model_name}/imu</topic>
    </sensor>
  </gazebo>
  <gazebo reference="lidar_link">
     <sensor name='gpu_lidar' type='gpu_lidar'>
      <pose>0 0 0 0 0 0</pose>
      <topic>model/${model_name}/sonar</topic>
      <update_rate>10</update_rate>
      <lidar>
        <scan>
          <horizontal>
            <samples>1</samples>
            <resolution>1</resolution>
            <min_angle>0</min_angle>
            <max_angle>0</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.08</min>
          <max>10.0</max>
          <resolution>0.02</resolution>
        </range>
      </lidar>
      <alwaysOn>1</alwaysOn>
      <visualize>true</visualize>
    </sensor>
    <sensor name="logical_camera" type="logical_camera">
      <topic>model/${model_name}/logical_camera</topic>
      <logical_camera>
        <near>0.55</near>
        <far>5.5</far>
        <horizontal_fov>1.3962634</horizontal_fov>
        <aspect_ratio>1</aspect_ratio>
      </logical_camera>
      <visualize>true</visualize>
      <always_on>true</always_on>
      <update_rate>10</update_rate>
    </sensor>
  </gazebo>
</robot>